auth-spike.md

```diff
@@ -0,0 +1,75 @@
+- The `/logout` endpoint will log you out.
+
+Caching is a pig, so if you want to try different gmail accounts, its easiest to just use an incognito window.
+
+Once logged in, your access token is stored in an encrypted cookie in the browser (http only, only sent via https, stored against this serves domain only). Until it expires or you log out, said access token is used to check your roles (i.e - do you have "admin"?) when you hit the `/` endpoint.
+
+I have _not_ implement refresh logic or anything like that, it's just a spike.
+
+## Setup
+
+Set the environment variables:
+
+| name | description |
+| ---- | ----------- |
+| ENCRYPTION_KEY | When running locally a default one will be generated. |
+| OAUTH_DOMAIN | Get from auth0 account/someone else. |
+| OAUTH_CLIENT_ID | Get from auth0 account/someone else. |
+| OAUTH_CLIENT_SECRET | Get from auth0 account/someone else. |
+| OAUTH_REDIRECT_URL | When runing locally, use http://localhost:3000/callback
+
+## Usage
+
+```python
+pipenv run ./app/server.py
+```
+
+Navigating here [http://localhost:3000/](http://localhost:3000/) to see the `/` endpoint.
+
+
+## How it works
+
+### Flow
+
+This server is using the standard `oidc` scope. So a typical "log in with google" approach, I'm not detailed that here but it's roughly:
+
+- this server uses its secrets (the env vars) to get an authentication url and `state` value from the auth provider (auth0).
+- user is authenticated via identity provider login (google in this case)
+- auth provider (auth0) redirects users response to this servers `/callback` endpoint, providing `token` and `state` values as url parameters.
+- this server checks the `state` it was originally given matches the `state` parameter given to the callback url (avoid CSRF attacks, i.e confirm incoming request actually is from the auth provider).
+- the `token` is permission from the user to request a scoped (to "oidc") `access_token` on their behalf - so we do.
+- the `access_token` is encrypted to a `user` cookie set against this servers domain.
+- the `access_token` in the cookie is used to check (per request) if a given (now `authenticated`) user has the `role` required to get the resource in question - this is `authorization`.
+
+### Rules
+
+Auth0 [rules](https://auth0.com/docs/customize/rules) allow you to extend the information returned from the `/userinfo` endpoint, so **after** authentication, but **before** said userinfo is returned to the app.
+
+In this case I've set a `rule` on an [https://auth0.com/](https://auth0.com/) account to add an `admin` role under the roles namespace (see configuration.ini) but only for authenticated google `@gsscogs.uk` users.
+
+_Note - the roles namespace defined in the apps cofiguration **must** match the same one used in the auth `rule`._ 
+
+In production the `rule`'s would likely be more nuanced than this (some variation of calling off to a database to get a given users roles) but the approach would/could be the same or similar, i.e
```

Comment:
As discussed, it'd be good to try and get the [Auth0 RBAC](https://auth0.com/docs/authorization/rbac) stuff up and running properly.